### PR TITLE
feat(io): fs io (remove/rename/mkdir/rmdir) + UNIX socket io (listen/accept/connect/send/recv/close)

### PR DESCRIPTION
**Pull request by khoek**
Bring more fs functions and Unix sockets to lean.

### CODE REVIEW

File: library/system/io_interface.lean [hunk 1]
```diff
@@ -40,15 +40,22 @@ class monad_io (m : Type → Type → Type) :=
 (fail     : Π e α, e → m e α)
 (iterate  : Π e α, α → (α → m e (option α)) → m e α)
 -- Primitive Types
+(socket   : Type)
```
Comment (cipher1024) [inline library/system/io_interface.lean]: Can we move that to `monad_io_net_system`?

**Comment by cipher1024** [comment]
Is this what you would use if you wanted to create a binding for a database (e.g. postgres)?

**Comment by khoek** [comment]
If the database server was running on the same machine, yes.

**Comment by cipher1024** [comment]
Perfect! I'll see if I can find some time to build database bindings. That could be useful for web apps and other popular application domains

File: library/system/io_interface.lean (hunk 1 above)
Comment (khoek) [reply library/system/io_interface.lean]: The only thing is that this mirrors the situation with core's `handle : Type`, which is also defined in `monad_io` but is used in the `fs.*` named IO functions.

Comment (cipher1024) [reply library/system/io_interface.lean]: I thought that might be your reason. I think the rationale for `handle` to be there is that it's used in `monad_io_file_system`, in `monad_io_serial` and `monad_io_process` and those classes don't need to depend on each other.

Comment (khoek) [reply library/system/io_interface.lean]: Ok sure :)

**Comment by khoek** [comment]
Now you could potentially write a database server in lean as well ;)

**Comment by cipher1024** [comment]
I'd love to see some test cases when you have a moment

**Comment by khoek** [comment]
@cipher1024 Test included, and it runs on windows now too.

**Comment by khoek** [comment]
If I used git right, that should be good.

**Comment by cipher1024** [comment]
I just saw the tests. They look great! Thank you for writing that!

**Comment by khoek** [comment]
I'm done with this one for now, too. I'll submit a separate pull for inet sockets, which are simple extension of what exists now.

**Comment by khoek** [comment]
As a side-effect, `leanpkg` won't break on bare windows anymore, which I think is a great plus

**Comment by khoek** [comment]
(It used to need an `mkdir` with `-p` semantics.)

**Comment by bryangingechen** [comment]
It turns out that this commit has been breaking the emscripten build. Do these features make sense in that context or would it make more sense for me to disable them?

**Comment by cipher1024** [comment]
Maybe they should be disabled or implemented otherwise. @agentultra, any way we can implement sockets with emscripten?

**Comment by bryangingechen** [comment]
It's looking like the other PR is building after I added: `|| defined(LEAN_EMSCRIPTEN))` in a few places. Let's continue the discussion [there](https://github.com/leanprover-community/lean/pull/17).

**Comment by agentultra** [comment]
I’m not too familiar with Emscriptem but unless it has a way to translate the Unix socket API to the Web Socket API then I think the answer would be no.

**Comment by bryangingechen** [comment]
Apparently it exists, though the only docs I could find on emscripten's socket-to-websocket functionality are hidden in [this open PR](https://github.com/emscripten-core/emscripten/blob/503c7889349c2f28e39e8b20a21dcdea71200c63/site/source/docs/porting/networking.rst#emulated-posix-tcp-sockets-over-websockets).

**Pull request merged**